keps/sig-node/20190925-enable-seccomp-by-default.md

```diff
@@ -0,0 +1,327 @@
+    SeccompProfileLocalhost  SeccompProfileType = "Localhost"
+    SeccompProfileKubernetes SeccompProfileType = "Kubernetes"
+)
+```
+
+Profiles will be passed as a serialised json object inside dockerOpt objects, in the same way that it is done currently for file based profiles:
+```
+jsonSeccomp, _ := json.Marshal(profile.Spec)
+return []dockerOpt{{"seccomp", string(jsonSeccomp), seccompProfileName}}, nil
+```
+_Needs confirmation as to whether this would also work for non-docker runtime implementations._
+
+
+Built-in profiles will act in the same way as custom profiles, therefore, trying to apply any of the audit built-in profiles in nodes which do not support `SCMP_ACT_LOG` (due to kernel or dependencies version) will cause containers to fail to start.
+
+
+
+
+### 3. Set `kubernetes/default-audit` as the default profile
+
+To define what seccomp profile a container will use, the following order of priority must be observed:
+
+1. Profile name set at container-level (TODO: update with seccomp GA field)
+1. Profile name set at pod-level (TODO: update with seccomp GA field)
+1. Default profile name set at PodSecurityPolicy (TODO: update with seccomp GA field)
+
+If none found, falls back to assessment to take place in the kubelet prior to creating the container:
+1. Is Libseccomp-golang higher than v0.9.1, and runC higher than runc 1.0-rc8 then use `kubernetes/default-audit`.  TODO: need to confirm what version of runC will support this, changes are already merged but a new release needs to happen.
+1. Fallback to `unconfined`.
+
+
+
+## Design Details
+
+### Test Plan
+
+Seccomp already has E2E tests, but the tests are guarded by the [Feature:Seccomp] tag and not run in the standard test suites.
+
+As part of seccomp [being marked GA](https://github.com/kubernetes/enhancements/pull/1148), the feature tag will be removed from the seccomp tests.
+
+New tests will be added covering profile selection logic described under [Implementation Details](#implementation-details). 
+
+
+
+### Upgrade / Downgrade Strategy
+
+No upgrade changes required. 
+
+Users can opt-out of this feature by setting the default seccomp profile to `unconfined` on a `PodSecurityPolicy` object. 
+
+
+### Version Skew Strategy
+
+This feature is based on kubelet changes only. 
+
+Older Kubelet versions won't be able set the default profile to `kubernetes/default-audit`, keeping the current setting of `unconfined` instead.
+
+The new built-in profiles will only be available from this version onwards. To use them, the pods must be scheduled on nodes in which the kubelet support them, otherwise they will fail to start.
+
+
+### Graduation Criteria
+
+##### Alpha
+
+- Make Built-in profiles available.
+- API changes to map Built-in profiles to an additional `Kubernetes` type.
+
+##### Alpha -> Beta Graduation
+
+- Gather feedback among community and developers.
+
+##### Beta -> GA Graduation
+
+- Enable audit-mode as default.
```

Comment:
Once the tooling is sufficiently mature, and with enough soak time, do you see us eventually flipping the default to the enforcing default profile?

Reply:
Yes, I am adding that now. My view is that the GA state of this KEP will be all built-in pods using `kubernetes/default`. The `kubernetes/default-audit` profile will be kept as default for user workloads.